apps/prairielearn/src/tests/workspaceDynamicFiles.test.ts

```diff
@@ -0,0 +1,173 @@
+import * as cheerio from 'cheerio';
+import { assert } from 'chai';
+import fetch from 'node-fetch';
+import * as fs from 'node:fs/promises';
+import { join } from 'node:path';
+
+import { config } from '../lib/config.js';
+import { initialize } from '../lib/workspace.js';
+import * as helperServer from './helperServer.js';
+
+import { loadSqlEquiv, queryRow, queryRows } from '@prairielearn/postgres';
+import { IdSchema, IssueSchema } from '../lib/db-types.js';
+
+const sql = loadSqlEquiv(import.meta.url);
+
+const siteUrl = 'http://localhost:' + config.serverPort;
+const baseUrl = siteUrl + '/pl';
+
+const originalConfig = { ...config };
+
+let workspaceId: string | undefined;
+let sourcePath: string;
+let variantId: string | undefined;
+
+async function checkFileContents(filename: string, expectedContents: string) {
+  const fileContents = await fs.readFile(join(sourcePath, filename), 'binary');
+  assert.equal(fileContents, expectedContents);
+}
+
+function createAndInitializeWorkspaceQuestion(qid: string) {
+  it('create workspace question variant', async () => {
+    const questionId = await queryRow(sql.select_test_question, { qid }, IdSchema);
+    const workspaceQuestionUrl = `${baseUrl}/course_instance/1/instructor/question/${questionId}/preview`;
+    const response = await fetch(workspaceQuestionUrl);
+
+    const $ = cheerio.load(await response.text());
+    const workspaceButton = $('a:contains("Open workspace")');
+    assert.lengthOf(workspaceButton, 1);
+
+    workspaceId = workspaceButton.attr('href')?.match('/pl/workspace/([0-9]+)')?.[1];
+    assert.isDefined(workspaceId);
+
+    variantId = $('.question-form input[name="__variant_id"]').attr('value');
+    assert.isDefined(variantId);
+  });
+
+  it('initialize workspace', async () => {
+    assert.isDefined(workspaceId);
+    // A workspace is typically initialized using sockets, but to simplify the
+    // testing environment, we will call the initialization process directly.
+    ({ sourcePath } = await initialize(workspaceId));
+    const directoryStats = await fs.stat(sourcePath);
+    assert.isTrue(directoryStats.isDirectory());
+  });
+}
```

Comment:
Rather than having a function that produces `it` test blocks, just make this a normal async function and call it when it's needed. We want to move away from this pattern where we have test cases magically produced by other function calls.

Reply:
Done in 696e5d1.